### DISCUSSION TITLE
Proposal: CallerTypeNameAttribute to provide nameof class/struct that is calling method.

### DISCUSSION THREAD

**Discussion by nathan130200** [dotnet/runtime#102724] [Ideas]
As the title says, the introduction of the attribute `[CallerTypeName]` which will provide the method with the name of the type that is invoking the current method, as well as `[CallerMemberName]`. Used especially in logging functions.

In asynchronous functions (context: async/await/task) the value will receive the original type that invoked this function. 

Example: 
```cs
static class Utilities
{

    void WriteToStdout([CallerTypeName] string typeName = default, [CallerMemberName] string memberName = default)
    {
        Console.WriteLine("[{0}::{1}]: invoked!", typeName, memberName);
    }

}
```

- `typeName` will be `nameof(Utilities)` same behaviour as `memberName` that will be `WriteToStdout`.

I had this problem while writing a micro logger to handle specific tasks with more logging details. And when I realized that I was either creating a gigantic work-around or creating scoped loggers that store the methodbase instance or method/type name. 

And as C# does not support expansive macros like C++, it is not possible to simply write a macro that calls all the necessary functions by passing `MethodBase.GetCurrentMethod()` in each of them. (it would be something like `#define INVOKE_LOG(LogInst, Msg) LogInst.WriteLine(Msg, MethodBase.GetCurrentMethod())`)

**Comment by AustinWise** [comment]
There is a discussion in the C# Language repo with a similar proposal: https://github.com/dotnet/csharplang/discussions/87
